// @glimmer/syntax/dist/commonjs/es5/lib/source/span-list.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.loc = loc;
exports.hasSpan = hasSpan;
exports.maybeLoc = maybeLoc;
exports.SpanList = void 0;

var _span2 = require("./span");

var __classPrivateFieldSet = void 0 && (void 0).__classPrivateFieldSet || function (receiver, privateMap, value) {
  if (!privateMap.has(receiver)) {
    throw new TypeError("attempted to set private field on non-instance");
  }

  privateMap.set(receiver, value);
  return value;
};

var __classPrivateFieldGet = void 0 && (void 0).__classPrivateFieldGet || function (receiver, privateMap) {
  if (!privateMap.has(receiver)) {
    throw new TypeError("attempted to get private field on non-instance");
  }

  return privateMap.get(receiver);
};

var _span;

var SpanList = /*#__PURE__*/function () {
  function SpanList(span) {
    if (span === void 0) {
      span = [];
    }

    _span.set(this, void 0);

    __classPrivateFieldSet(this, _span, span);
  }

  SpanList.range = function range(span, fallback) {
    if (fallback === void 0) {
      fallback = _span2.SourceSpan.NON_EXISTENT;
    }

    return new SpanList(span.map(loc)).getRangeOffset(fallback);
  };

  var _proto = SpanList.prototype;

  _proto.add = function add(offset) {
    __classPrivateFieldGet(this, _span).push(offset);
  };

  _proto.getRangeOffset = function getRangeOffset(fallback) {
    if (__classPrivateFieldGet(this, _span).length === 0) {
      return fallback;
    } else {
      var first = __classPrivateFieldGet(this, _span)[0];

      var last = __classPrivateFieldGet(this, _span)[__classPrivateFieldGet(this, _span).length - 1];

      return first.extend(last);
    }
  };

  return SpanList;
}();

exports.SpanList = SpanList;
_span = new WeakMap();

function loc(span) {
  if (Array.isArray(span)) {
    var first = span[0];
    var last = span[span.length - 1];
    return loc(first).extend(loc(last));
  } else if (span instanceof _span2.SourceSpan) {
    return span;
  } else {
    return span.loc;
  }
}

function hasSpan(span) {
  if (Array.isArray(span) && span.length === 0) {
    return false;
  }

  return true;
}

function maybeLoc(location, fallback) {
  if (hasSpan(location)) {
    return loc(location);
  } else {
    return fallback;
  }
}
