// @glimmer/syntax/dist/commonjs/es5/lib/v2-a/objects/expr.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.isLiteral = isLiteral;
exports.InterpolateExpression = exports.CallExpression = exports.PathExpression = exports.LiteralExpression = void 0;

var _slice = require("../../source/slice");

var _node = require("./node");

function _inheritsLoose(subClass, superClass) {
  subClass.prototype = Object.create(superClass.prototype);
  subClass.prototype.constructor = subClass;
  subClass.__proto__ = superClass;
}

/**
 * Corresponds to a Handlebars literal.
 *
 * @see {LiteralValue}
 */
var LiteralExpression = /*#__PURE__*/function (_node$fields) {
  _inheritsLoose(LiteralExpression, _node$fields);

  function LiteralExpression() {
    return _node$fields.apply(this, arguments) || this;
  }

  var _proto = LiteralExpression.prototype;

  _proto.toSlice = function toSlice() {
    return new _slice.SourceSlice({
      loc: this.loc,
      chars: this.value
    });
  };

  return LiteralExpression;
}((0, _node.node)('Literal').fields());
/**
 * Returns true if an input {@see ExpressionNode} is a literal.
 */


exports.LiteralExpression = LiteralExpression;

function isLiteral(node, kind) {
  if (node.type === 'Literal') {
    if (kind === undefined) {
      return true;
    } else if (kind === 'null') {
      return node.value === null;
    } else {
      return typeof node.value === kind;
    }
  } else {
    return false;
  }
}
/**
 * Corresponds to a path in expression position.
 *
 * ```hbs
 * this
 * this.x
 * @x
 * @x.y
 * x
 * x.y
 * ```
 */


var PathExpression = /*#__PURE__*/function (_node$fields2) {
  _inheritsLoose(PathExpression, _node$fields2);

  function PathExpression() {
    return _node$fields2.apply(this, arguments) || this;
  }

  return PathExpression;
}((0, _node.node)('Path').fields());
/**
 * Corresponds to a parenthesized call expression.
 *
 * ```hbs
 * (x)
 * (x.y)
 * (x y)
 * (x.y z)
 * ```
 */


exports.PathExpression = PathExpression;

var CallExpression = /*#__PURE__*/function (_node$fields3) {
  _inheritsLoose(CallExpression, _node$fields3);

  function CallExpression() {
    return _node$fields3.apply(this, arguments) || this;
  }

  return CallExpression;
}((0, _node.node)('Call').fields());
/**
 * Corresponds to an interpolation in attribute value position.
 *
 * ```hbs
 * <a href="{{url}}.html"
 * ```
 */


exports.CallExpression = CallExpression;

var InterpolateExpression = /*#__PURE__*/function (_node$fields4) {
  _inheritsLoose(InterpolateExpression, _node$fields4);

  function InterpolateExpression() {
    return _node$fields4.apply(this, arguments) || this;
  }

  return InterpolateExpression;
}((0, _node.node)('Interpolate').fields());

exports.InterpolateExpression = InterpolateExpression;
