// @glimmer/util/dist/commonjs/es2017/lib/collections.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.dict = dict;
exports.isDict = isDict;
exports.isObject = isObject;
exports.NonemptyStackImpl = exports.StackImpl = exports.DictSet = void 0;

var _guid = require("./guid");

function dict() {
  return Object.create(null);
}

function isDict(u) {
  return u !== null && u !== undefined;
}

function isObject(u) {
  return typeof u === 'object' && u !== null;
}

class DictSet {
  constructor() {
    this.dict = dict();
  }

  add(obj) {
    if (typeof obj === 'string') this.dict[obj] = obj;else this.dict[(0, _guid.ensureGuid)(obj)] = obj;
    return this;
  }

  delete(obj) {
    if (typeof obj === 'string') delete this.dict[obj];else if (obj._guid) delete this.dict[obj._guid];
  }

}

exports.DictSet = DictSet;

class StackImpl {
  constructor(values = []) {
    this.current = null;
    this.stack = values;
  }

  get size() {
    return this.stack.length;
  }

  push(item) {
    this.current = item;
    this.stack.push(item);
  }

  pop() {
    let item = this.stack.pop();
    let len = this.stack.length;
    this.current = len === 0 ? null : this.stack[len - 1];
    return item === undefined ? null : item;
  }

  nth(from) {
    let len = this.stack.length;
    return len < from ? null : this.stack[len - from];
  }

  isEmpty() {
    return this.stack.length === 0;
  }

  toArray() {
    return this.stack;
  }

}

exports.StackImpl = StackImpl;

class NonemptyStackImpl {
  constructor(values) {
    this.stack = values;
    this.current = values[values.length - 1];
  }

  get size() {
    return this.stack.length;
  }

  push(item) {
    this.current = item;
    this.stack.push(item);
  }

  pop() {
    if (this.stack.length === 1) {
      throw new Error(`cannot pop the last element of a NonemptyStack`);
    }

    let item = this.stack.pop();
    let len = this.stack.length;
    this.current = this.stack[len - 1];
    return item;
  }

  nth(from) {
    let len = this.stack.length;
    return from >= len ? null : this.stack[from];
  }

  nthBack(from) {
    let len = this.stack.length;
    return len < from ? null : this.stack[len - from];
  }

  toArray() {
    return this.stack;
  }

}

exports.NonemptyStackImpl = NonemptyStackImpl;
