// @glimmer/syntax/dist/modules/es2017/lib/source/span-list.js
var __classPrivateFieldSet = this && this.__classPrivateFieldSet || function (receiver, privateMap, value) {
  if (!privateMap.has(receiver)) {
    throw new TypeError("attempted to set private field on non-instance");
  }

  privateMap.set(receiver, value);
  return value;
};

var __classPrivateFieldGet = this && this.__classPrivateFieldGet || function (receiver, privateMap) {
  if (!privateMap.has(receiver)) {
    throw new TypeError("attempted to get private field on non-instance");
  }

  return privateMap.get(receiver);
};

var _span;

import { SourceSpan } from './span';
export class SpanList {
  constructor(span = []) {
    _span.set(this, void 0);

    __classPrivateFieldSet(this, _span, span);
  }

  static range(span, fallback = SourceSpan.NON_EXISTENT) {
    return new SpanList(span.map(loc)).getRangeOffset(fallback);
  }

  add(offset) {
    __classPrivateFieldGet(this, _span).push(offset);
  }

  getRangeOffset(fallback) {
    if (__classPrivateFieldGet(this, _span).length === 0) {
      return fallback;
    } else {
      let first = __classPrivateFieldGet(this, _span)[0];

      let last = __classPrivateFieldGet(this, _span)[__classPrivateFieldGet(this, _span).length - 1];

      return first.extend(last);
    }
  }

}
_span = new WeakMap();
export function loc(span) {
  if (Array.isArray(span)) {
    let first = span[0];
    let last = span[span.length - 1];
    return loc(first).extend(loc(last));
  } else if (span instanceof SourceSpan) {
    return span;
  } else {
    return span.loc;
  }
}
export function hasSpan(span) {
  if (Array.isArray(span) && span.length === 0) {
    return false;
  }

  return true;
}
export function maybeLoc(location, fallback) {
  if (hasSpan(location)) {
    return loc(location);
  } else {
    return fallback;
  }
}
