// @glimmer/syntax/dist/modules/es2017/lib/traversal/path.js
import RootTransformScope from './scope';
export default class WalkerPath {
  constructor(node, parent = null, parentKey = null) {
    this.node = node;
    this.parent = parent;
    this.parentKey = parentKey;
    this.scope = parent ? parent.scope.child(node) : new RootTransformScope(node); // Consume in scope values

    if (node.type === 'PathExpression') {
      this.scope.useLocal(node);
    }

    if (node.type === 'ElementNode') {
      this.scope.useLocal(node);
      node.children.forEach(node => this.scope.useLocal(node));
    }
  }

  get parentNode() {
    return this.parent ? this.parent.node : null;
  }

  parents() {
    return {
      [Symbol.iterator]: () => {
        return new PathParentsIterator(this);
      }
    };
  }

}

class PathParentsIterator {
  constructor(path) {
    this.path = path;
  }

  next() {
    if (this.path.parent) {
      this.path = this.path.parent;
      return {
        done: false,
        value: this.path
      };
    } else {
      return {
        done: true,
        value: null
      };
    }
  }

}
